### PR TITLE
Revision of VB arrays topic

### LINKED ISSUE #3760: Arrays in Visual Basic - High PV low CSAT - code example issues
# Arrays in Visual Basic - High PV low CSAT - code example issues
This topic: [Arrays in Visual Basic](https://docs.microsoft.com/en-us/dotnet/visual-basic/programming-guide/language-features/arrays/)

Very high page views (23K in one month), but low customer satisfaction ratings and broken code therefore suggesting a high priority.

Requires further inspection, but the obvious issues called out in the voting feedback are:

- Issues in code examples.
- Asks for info/examples on splitting and joining arrays.

### PR DESCRIPTION
**Pull request by rpetrusha**
# Revision of VB arrays topic

## Summary

Revises the VB arrays topic to remove non-compilable code and address customer issues, especially lack of clarity and absence of examples.

Fixes #3760 

## Suggested Reviewers

@BillWagner

### CODE REVIEW

File: docs/visual-basic/programming-guide/language-features/arrays/index.md [hunk 1]
```diff
@@ -14,257 +12,311 @@ helpviewer_keywords:
...
+
+It is possible individual variables to store each of our data items. For example, if our application analyzes student grades, we can use a separate variable for each student's grade, such as `englishGrade1`, `englishGrade2`, etc. This approach has three major limitations:
+- We have to know at design time exactly how many grades we have to handle.
+- Handling large numbers of grades quickly become unwieldy. This in turn makes an application much more likely to have serious bugs.
+- It is difficult to maintain. Each new grade that we add requires that the application be modified, recompiled, and redeployed.  
+ 
+ By using an array, you can refer to these related values by the same name, and use a number that’s called an *index* or *subscript* to identify an individual element based on its position in the array. The indexes of an array range from 0 to one less than the total number of elements in the array. You can work with the array as a unit, and the ability to iterate its elements frees you from needing to know exactly how many elements it contains at design time.
```
Comment (BillWagner) [inline docs/visual-basic/programming-guide/language-features/arrays/index.md]: nit: indices or indexes?  (I'm not sure which is preferred by our style guide). There are several occurrences of "indexes" in this article. Whichever is correct, let's be consistent.

Comment (BillWagner) [inline docs/visual-basic/programming-guide/language-features/arrays/index.md]: I'd add one sentence that explains that array sizes are specified as the max index, not the size. That way, this important point is introduced before the examples.

File: docs/visual-basic/programming-guide/language-features/arrays/index.md [hunk 2]
```diff
@@ -14,257 +12,311 @@ helpviewer_keywords:
...
-  
- [!code-vb[VbVbalrArrays#3](../../../../../samples/snippets/visualbasic/VS_Snippets_VBCSharp/VbVbalrArrays/VB/Class1.vb#3)]  
-  
- You can refer to the array as a whole by using just the array variable name without indexes.  
-  
- The array `students` in the preceding example uses one index and is said to be one-dimensional. An array that uses more than one index or subscript is called multidimensional. For more information, see the rest of this topic and [Array Dimensions in Visual Basic](../../../../visual-basic/programming-guide/language-features/arrays/array-dimensions.md).  
-  
-##  <a name="BKMK_CreatingAnArray"></a> Creating an Array  
- You can define the size of an array several ways. You can supply the size when the array is declared, as the following example shows.  
+ The `students` array in the preceding example is a one-dimensional array because it uses one index. An array that uses more than one index or subscript is called multidimensional. For more information, see the rest of this article and [Array Dimensions in Visual Basic](../../../../visual-basic/programming-guide/language-features/arrays/array-dimensions.md).
```
Comment (BillWagner) [inline docs/visual-basic/programming-guide/language-features/arrays/index.md]: should multidimensional be hyphenated?

File: docs/visual-basic/programming-guide/language-features/arrays/index.md [hunk 3]
```diff
@@ -14,257 +12,311 @@ helpviewer_keywords:
...
   
--   Your code interacts with an application programming interface (API) that either requires you to pass a zero-length array to one or more procedures or returns a zero-length array from one or more procedures.  
+ The following example uses a [For Each...Next Statement](../../../../visual-basic/language-reference/statements/for-each-next-statement.md)to iterate through a one-dimensional array and a two-dimensional array.  
   
-##  <a name="BKMK_ArraySize"></a> Array Size  
- The size of an array is the product of the lengths of all its dimensions. It represents the total number of elements currently contained in the array.  
+ [!code-vb[iterate-for-each-next](../../../../../samples/snippets/visualbasic/programming-guide/language-features/arrays/iterate-for-each-next.vb)]  
   
- The following example declares a three-dimensional array.  
+## Array Size
```
Comment (BillWagner) [inline docs/visual-basic/programming-guide/language-features/arrays/index.md]: I would mention somewhere in this section that none of this applies to jagged arrays (also referred to as "arrays of arrays")

File: docs/visual-basic/programming-guide/language-features/arrays/index.md [hunk 4]
```diff
@@ -14,257 +12,311 @@ helpviewer_keywords:
...
   
-##  <a name="BKMK_ArrayTypes"></a> Array Types and Other Types  
- Every array has a data type, but it differs from the data type of its elements. There is no single data type for all arrays. Instead, the data type of an array is determined by the number of dimensions, or *rank*, of the array, and the data type of the elements in the array. Two array variables are considered to be of the same data type only when they have the same rank and their elements have the same data type. The lengths of the dimensions in an array do not influence the array data type.  
+ Every array inherits from the <xref:System.Array?displayProperty=nameWithType> class, and you can declare a variable to be of type `Array`, but you cannot create an array of type `Array`. For example, although the following code declares the `arr` variable to be of type `Array` and calls the <xref:System.Array.CreateInstance%2A?displayProperty=nameWithType> method to instantiate the array, the array's type proves to be Object[].
```
Comment (BillWagner) [inline docs/visual-basic/programming-guide/language-features/arrays/index.md]: I'd consider removing the example: it demonstrates what we recommend *not* to do.

**Review by BillWagner — APPROVED** [review]
Overall this looks really good.

I had a few suggestions and one question. You can decide what to do with each, and :shipit: when you are ready.

File: docs/visual-basic/programming-guide/language-features/arrays/index.md (hunk 1 above)
Comment (rpetrusha) [reply docs/visual-basic/programming-guide/language-features/arrays/index.md]: @BillWagner, indexes is preferred.

Comment (BillWagner) [reply docs/visual-basic/programming-guide/language-features/arrays/index.md]: I'll remember that.

File: docs/visual-basic/programming-guide/language-features/arrays/index.md (hunk 2 above)
Comment (rpetrusha) [reply docs/visual-basic/programming-guide/language-features/arrays/index.md]: No, it shouldn't. I was inconsistent, so I've removed the hyphen.

File: docs/visual-basic/programming-guide/language-features/arrays/index.md (hunk 3 above)
Comment (rpetrusha) [reply docs/visual-basic/programming-guide/language-features/arrays/index.md]: @BillWagner, yes. Your comment points to a weakness of the existing topic -- I didn't discuss determining the size of a jagged array, so I'll add that and an example to the jagged arrays section. When I do that (hopefully later today, but maybe not until tomorrow) I'd like you to take a look at the latest round of revisions.

**Comment by rpetrusha** [comment]
@BillWagner, I've addressed your comments, discussed the size of jagged arrays, and added an example.

**Comment by BillWagner** [comment]
@rpetrusha I read through the changes you made, and this is great.  Let's :shipit: when you are ready.

**Pull request merged**